### PR TITLE
feat(#9431): always aggregate and store targets

### PR DESCRIPTION
**Pull request by dianabarsan**
# Description

Refactors the calculation and aggregation of targets and recalculation of tasks from only being triggered when visiting certain pages to being calculated automatically when the state changes or documents are updated. 
To account for heavy recalculation cycles when multiple documents are downloaded or created, there is a 1s debounce between receiving a change and emissions recalculations trigger. 

#9431

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9431-always-aggregate/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9431-always-aggregate/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:9431-always-aggregate/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by dianabarsan** [comment]
Hi @lorerod 

I still want to add one e2e test to check how targets get generated now, but I'd love to have some feedback on what you're observing when running this. 

also hi @garethbowen 

Would love another set of eyes on this.

File: webapp/src/ts/services/rules-engine.service.ts
Comment (garethbowen) [inline webapp/src/ts/services/rules-engine.service.ts:251]: I think this callback handler is complicated enough that it deserves to be its own named function...

**Review by garethbowen — CHANGES_REQUESTED** [review]
Some suggestions, one bug (I think).

File: shared-libs/rules-engine/src/pouchdb-provider.js
Comment (dianabarsan) [reply shared-libs/rules-engine/src/pouchdb-provider.js:87]: Yea, true, I'll make the change

Comment (dianabarsan) [reply shared-libs/rules-engine/src/pouchdb-provider.js:87]: There are so many 409s on this doc reported as feedback docs it's insane. 
This PR should reduce that by throttling setting contacts dirty.

File: shared-libs/rules-engine/src/provider-wireup.js
Comment (dianabarsan) [reply shared-libs/rules-engine/src/provider-wireup.js:80]: This is the de-facto way of how this library is structured, unfortunately. I can change it for this function.

File: tests/e2e/default/purge/purge.wdio-spec.js
Comment (dianabarsan) [reply tests/e2e/default/purge/purge.wdio-spec.js:108]: Because we add 500 docs :| 
But I think this can be removed, because I turned off the rules-engine for this test.

File: tests/e2e/default/transitions/create-user-for-contacts.replace-user.wdio-spec.js
Comment (dianabarsan) [reply tests/e2e/default/transitions/create-user-for-contacts.replace-user.wdio-spec.js:275]: This is only the case for this user-replace test where webapp actually logs the user out after sync.

File: tests/e2e/default/transitions/create-user-for-contacts.replace-user.wdio-spec.js
Comment (dianabarsan) [reply tests/e2e/default/transitions/create-user-for-contacts.replace-user.wdio-spec.js:835]: It's not unrelated. 
Because we recalculate in the background now, sync will happen outside of the test, and it generated feedback docs once the user became "disabled" server-side.

File: webapp/src/ts/services/rules-engine.service.ts
Comment (dianabarsan) [reply webapp/src/ts/services/rules-engine.service.ts:251]: Agree!

File: tests/page-objects/default/login/login.wdio.page.js
Comment (dianabarsan) [reply tests/page-objects/default/login/login.wdio.page.js:48]: I've had tests fail because the snackbar was up and got the click. The behavior of the app changes slightly because of the doc changes that happen in the background now.

File: tests/utils/index.js
Comment (dianabarsan) [reply tests/utils/index.js:1642]: haha, I had a slur of e2e tests failing. At some point some test couldn't select a form because it was obscured by other forms listed in the popup. So I changed how test cleanup is handled to only keep default forms around and delete ephemeral ones. 

Do you think this is important enough to move to another PR?

File: webapp/src/ts/services/rules-engine.service.ts
Comment (dianabarsan) [reply webapp/src/ts/services/rules-engine.service.ts:129]: Oh! Thanks so much for spotting this!!

File: webapp/src/ts/services/rules-engine.service.ts
Comment (dianabarsan) [reply webapp/src/ts/services/rules-engine.service.ts:138]: Good suggestion!

File: shared-libs/rules-engine/src/target-state.js
Comment (dianabarsan) [reply shared-libs/rules-engine/src/target-state.js:139]: The end is not the end of the month, it's last millisecond before the 1st of the month (example: 1730411999999 - this is localized to my TZ though)

File: tests/page-objects/default/common/common.wdio.page.js
Comment (dianabarsan) [reply tests/page-objects/default/common/common.wdio.page.js:30]: Sync patterns change  a bit. The old selector could have hit the snackbar (it uses the same text) disappearing instead of the menu item. so this is still correct.

**Review by garethbowen — APPROVED** [review]
Looks great. Approved to unblock. Feel free to request another review if you'd like!

**Comment by lorerod** [comment]
**Performance Testing Steps with Default Config Using a Local Instance and Chrome desktop:**
**Data Setup:**
Created 40 clinics, each with 10 patients.
**_The places, patients and reports were created using `/_bulk_docs`_

**Steps:**
1. Logged in with an offline user.
2. Created 1 pregnancy report for each patient, leading to 400 pregnancy reports.**
3. Opened the hamburger menu and initiated the sync process, started a stopwatch, and measured the time it took for the sync process to finish.

| Step | Average using branch 9431-always-aggregate  | Average using branch master  |
| ------------ | ------------ | ------------ |
| Sync after creating reports  | 0:02:08  | 0:00:17  |

**Steps:**
1. Logged in with an offline user.
2. Created 1 pregnancy report for each patient, leading to 400 pregnancy reports.**
3. Opened the hamburger menu and initiated the sync process.
5. After starting the sync, I closed the hamburger menu and didn't wait for the sync process to finish. 
6. Navigated to the Targets tab, started a stopwatch, and measured the time it took for the page to load, showing the 400 pregnancies.
7. Repeated the same process (stopwatch and time measurement) for the Tasks tab, Reports tab, and Contacts tab, recording the time it took for each tab to load completely in the order listed.

| Step | Average using branch 9431-always-aggregate  | Average using branch master  |
| ------------ | ------------ | ------------ |
| Loading Targets tab | 0:00:14  | 0:00:21  |
| Loading Tasks Tab.  | 0:00:14  |  0:00:18 |
| Loading Reports tab  | 0:00:05  | 0:00:05  |
| Loading Contacts Tab.  | 0:00:08  | 0:00:07  |

**Steps:**
1. Logged in with an offline user.
2. Created 1 pregnancy report for each patient, leading to 400 pregnancy reports.**
3. Opened the hamburger menu and initiated the sync process.
4. After starting the sync, I closed the hamburger menu and didn't wait for the sync process to finish.
5. Navigated to the Contacts tab, started a stopwatch, and measured the time it took for the page to load.

| Step | Average using branch 9431-always-aggregate  | Average using branch master  |
| ------------ | ------------ | ------------ |
| Loading Contacts Tab.  | 0:00:08  | 0:00:07  |

The `9431-always-aggregate` branch introduces a significantly longer sync time but shows faster load times for the Targets and Tasks tabs. Other tabs and scenarios show no relevant differences.

@dianabarsan, do these numbers look correct to you? Let me know if there are other workflows you'd like me to test, or if you'd prefer I rerun any tests to confirm the increased sync time.

File: shared-libs/rules-engine/src/target-state.js
Comment (kennsippell) [inline shared-libs/rules-engine/src/target-state.js:37]: needed?

File: webapp/src/ts/services/rules-engine.service.ts
Comment (kennsippell) [inline webapp/src/ts/services/rules-engine.service.ts:153]: :partying_face:

File: shared-libs/rules-engine/src/provider-wireup.js
Comment (kennsippell) [inline shared-libs/rules-engine/src/provider-wireup.js:313]: i'm a bit unclear when `handleIntervalTurnover` is now necessary. since we write the target doc as soon as data changes

**Comment by kennsippell** [comment]
And notes during testing:

* Deleting a doc through UI triggers a recalculation, but deleting it through fauxton doesn't cause recalculation. The value only updates after 7 days.
* Tried to make _rev bump needlessly but could not.
* Tried syncing a few thousand docs to see how many times targets are recalculated - behavior seemed good.
* Tried to create and sync a doc + close app before target recalculates + bump clock ahead to next month + login fresh from another phone. This made an incorrect score in target doc.

Can I ask about the expected behavior for telemetry events? 

* This is adding `rules-engine:emissions:all-contacts` to measure performance of these background recalculations? Did you consider a name like `rules-engine:background:all-contacts` or alike? Might be more intuitive cause nobody wants to know what emissions are.
* What is the expectation for the `dirty-contacts` metric after this? Let's say I sync 100 reports which change data for 100 contacts (100 dirty contacts). If I wait for the sync, wait for target doc refresh, when navigate to targets tab - should I see zero dirty contacts in telemetry?

File: webapp/src/ts/services/rules-engine.service.ts
Comment (kennsippell) [inline webapp/src/ts/services/rules-engine.service.ts:358]: I think we should keep emissions as a word fore core devs whenever possible.  Did you consider a name like `rules-engine:background:all-contacts` or `rules-engine:proactive:all-contacts` or alike? Might be more intuitive. Nobody wants to know what emissions are.

**Comment by dianabarsan** [comment]
Thank you @lorerod for the comprehensive testing!! 

Indeed the sync time increase is expected because immediately after creating pregnancies, all tasks for each pregnancy are calculated. If you would open the tasks tab immediately after sync, you would observe the difference in time in loading, and on the next sync, after opening the tasks tab, you would see the difference in sync time, when this branch would be fast and master would be slow for both. 

Excellent testing and I'm really happy with the numbers!

File: shared-libs/rules-engine/src/index.js
Comment (dianabarsan) [reply shared-libs/rules-engine/src/index.js:44]: The naming already comes from another disambiguation, that Gareth requested. I think your comment applies to many interfaces in this library. Suggest a more appropriate name.

File: shared-libs/rules-engine/src/provider-wireup.js
Comment (dianabarsan) [reply shared-libs/rules-engine/src/provider-wireup.js:313]: It's not needed anymore, indeed. Gareth and I discussed it already, and we decided that due to the urgency on this, we would revisit it, and further cleanup of the library in a next iteration.

**Comment by dianabarsan** [comment]
> but deleting through fauxton + sync doesn't cause recalculation. Navigating to the target tab shows an incorrect value .

This has always been the case, because of how fauxton creates the deletion. Never delete through Fauxton. You can delete through the UI as an admin. 

> Tried to create and sync a doc + close app before target recalculates + bump clock ahead to next month + login fresh from another phone. I can see invalid scores.

This is also what happens when you close your computer before google docs saves your document to the cloud. 

> What is the expectation for the dirty-contacts metric after this? Let's say I sync 100 reports which change data for 100 contacts (100 dirty contacts). If I wait for the sync, wait for target doc refresh, when navigate to targets tab - should I see zero dirty contacts in telemetry?

`dirty-contacts` is another thing that must be changed in the next iteration. I think it only becomes relevant for the change watcher now, where I haven't added it, because all the other "fetch" tasks wait for the change watcher to finish calculating.

File: shared-libs/rules-engine/src/provider-wireup.js
Comment (dianabarsan) [reply shared-libs/rules-engine/src/provider-wireup.js:313]: There might be edge cases where this is still useful - ex: the user closes their app before saving the target doc and only opening the app in the old interval.

This was a talking point, about updating what `now` means and actually making it possible to completely recalculate targets within another interval, but the conclusion was that this change was not straightforward (decisions need to be made, like: do we keep a "now" that means "this emission should always be counted unless the date is before the thing that emits was created?" or is `now` a range?). 

I think we should revisit this once we can accurately recalculate the target counts for any interval.

**Comment by dianabarsan** [comment]
I'm not seeing any strong objections to this. Since time is pressing, I'm inclined to just merge this as is, and backport it to 4.13.x to have this released by the end of the week. 

I'll give this until tomorrow morning (~14 hours) for strong objections and then proceed.

**Comment by kennsippell** [comment]
Yes please merge. So this is going out in 4.13 official and not as an FR?

**Comment by dianabarsan** [comment]
4.13 official.

**Pull request merged**
